File: ntfy/home-assistant/web/src/api/devices.ts
import { api, createEventSource } from '../api.ts'

export interface Device {
  device_id: string
  ble_mac: string | null
  status: 'pending_claim' | 'pending_status' | 'active' | 'failed'
  created_at: number
  claimed_at: number | null
}

export interface BleDevice {
  address: string
  name: string
  rssi: number
}

export interface PairingResult {
  device_id: string
  name: string
  status: string
  qr_payload: string
}

export interface DeviceClaimedEvent {
  type: 'device_claimed'
  device_id: string
}

export async function listDevices(): Promise<Device[]> {
  const res = await api('/api/devices')
  return res.json()
}

export async function startBleScan(): Promise<void> {
  const res = await api('/api/devices/ble-scan/start', { method: 'POST' })
  if (!res.ok) {
    const data = await res.json().catch(() => ({}))
    throw new Error(data.detail || 'Failed to start BLE scan')
  }
}

export async function pairDevice(
  bleAddress: string,
  name: string,
  ssid: string,
  password: string,
): Promise<PairingResult> {
  const res = await api('/api/devices/pair', {
    method: 'POST',
    body: new URLSearchParams({
      ble_address: bleAddress,
      name,
      ssid,
      password,
    }),
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  })
  const data = await res.json()
  if (!res.ok) {
    throw new Error(data.detail || 'Failed to start pairing')
  }
  return data
}

export function subscribeBleScan(
  onDevice: (device: BleDevice) => void,
): EventSource {
  const es = createEventSource('/api/devices/ble-scan')
  es.onmessage = (e) => {
    try {
      onDevice(JSON.parse(e.data))
    } catch {
      // ignore malformed lines
    }
  }
  es.onerror = () => {
    es.close()
  }
  return es
}

export function subscribeDeviceEvents(
  onEvent: (event: DeviceClaimedEvent) => void,
): EventSource {
  const es = createEventSource('/api/messages/stream')
  es.onmessage = (e) => {
    try {
      const data = JSON.parse(e.data)
      if (data.type === 'device_claimed') {
        onEvent(data as DeviceClaimedEvent)
      }
    } catch {
      // ignore malformed lines
    }
  }
  es.onerror = () => {
    es.close()
  }
  return es
}

File: ntfy/home-assistant/server/src/home_assistant/api/devices.py
"""Device onboarding API for the web and Android frontends."""
import asyncio
import json
import time
from typing import Annotated

from fastapi import APIRouter, Form, HTTPException, Request
from fastapi.responses import StreamingResponse

from .. import auth, config
from ..devices import ble_scanner, qr_display, security as device_security
from ..devices import db as devices_db

router = APIRouter()


def _lan_ip() -> str:
    import socket
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    except OSError:
        return "127.0.0.1"
    finally:
        s.close()


@router.get("/devices")
def list_devices(current_user: auth.CurrentUser):
    return devices_db.list_devices()


@router.post("/devices/ble-scan/start")
async def start_ble_scan(request: Request, current_user: auth.CurrentUser):
    scanner = request.app.state.ble_scanner
    await scanner.start()
    return {"ok": True}


@router.get("/devices/ble-scan")
async def ble_scan_stream(request: Request, current_user: auth.CurrentUser):
    scanner = request.app.state.ble_scanner
    queue: asyncio.Queue[ble_scanner.BleDevice] = asyncio.Queue()

    def _on_device(device: ble_scanner.BleDevice) -> None:
        try:
            queue.put_nowait(device)
        except asyncio.QueueFull:
            pass

    scanner.on_device(_on_device)

    async def event_generator():
        while True:
            device = await queue.get()
            payload = {
                "address": device.address,
                "name": device.name,
                "rssi": device.rssi,
            }
            yield f"data: {json.dumps(payload)}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
    )


@router.post("/devices/pair")
def pair_device(
    current_user: auth.CurrentUser,
    csrf: auth.CsrfRequired,
    ble_address: Annotated[str, Form()],
    name: Annotated[str, Form()],
    ssid: Annotated[str, Form()],
    password: Annotated[str, Form()],
):
    """Generate credentials for the selected BLE device and show the QR code."""
    device_id = device_security.generate_device_id()
    token = device_security.generate_access_token()
    claim_key = device_security.generate_claim_key()
    token_salt, token_hash = device_security.hash_secret(token)
    claim_salt, claim_hash = device_security.hash_secret(claim_key)

    devices_db.add_device(
        device_id=device_id,
        token=token,
        token_salt=token_salt,
        token_hash=token_hash,
        claim_key=claim_key,
        claim_salt=claim_salt,
        claim_hash=claim_hash,
        ble_mac=ble_address,
    )

    server_host = _lan_ip()
    payload = qr_display.show_pairing_qr(
        claim_key=claim_key,
        ssid=ssid,
        password=password,
        server_host=server_host,
        serial_port=config.MSU_SCREEN_PORT or None,
    )

    return {
        "device_id": device_id,
        "name": name,
        "status": "pending_claim",
        "qr_payload": payload,
    }
